lib/Semantics/check-case.cpp

```diff
@@ -0,0 +1,244 @@
+        common::visitors{
+            [&](const std::list<parser::CaseValueRange> &ranges) {
+              for (const auto &range : ranges) {
+                cases_.emplace_back(stmt);
+                cases_.back().ComputeBounds(*this, range);
+              }
+            },
+            [&](const parser::Default &) { cases_.emplace_front(stmt); },
+        },
+        selector.u);
+  }
+
+  std::optional<Value> GetValue(const parser::CaseValue &caseValue) {
+    const auto &expr{caseValue.thing.thing.value()};
+    auto *x{expr.typedExpr.get()};
+    if (x && x->v) { // C1147
+      auto type{x->v->GetType()};
+      if (type && type->category() == caseExprType_.category() &&
+          (type->category() != TypeCategory::Character ||
+              type->kind() == caseExprType_.kind())) {
+        if (auto converted{evaluate::Fold(context_.foldingContext(),
+                evaluate::ConvertToType(T::GetType(), common::Clone(*x->v)))}) {
+          x->v = std::move(converted);
+          if (auto value{evaluate::GetScalarConstantValue<T>(*x->v)}) {
+            return *value;
+          }
+        }
+        context_.Say(
+            expr.source, "CASE value must be a constant scalar"_err_en_US);
+      } else {
+        std::string typeStr{type ? type->AsFortran() : "typeless"s};
+        context_.Say(expr.source,
+            "CASE value has type '%s' which is not compatible with the SELECT CASE expression's type '%s'"_err_en_US,
+            typeStr, caseExprType_.AsFortran());
+      }
+      x->v.reset(); // prevent repeated message
+      hasErrors_ = true;
+    }
+    return std::nullopt;
+  }
+
+  struct Case {
+    explicit Case(const parser::Statement<parser::CaseStmt> &s) : stmt{s} {}
+    bool IsDefault() const { return !lower && !upper; }
+    void ComputeBounds(
+        CaseValues &values, const parser::CaseValueRange &range) {
+      std::visit(
+          common::visitors{
+              [&](const parser::CaseValue &x) {
+                lower = values.GetValue(x);
+                upper = lower;
+              },
+              [&](const parser::CaseValueRange::Range &x) {
+                if (x.lower) {
+                  lower = values.GetValue(*x.lower);
+                }
+                if (x.upper) {
+                  upper = values.GetValue(*x.upper);
+                }
+              },
+          },
+          range.u);
+    }
+    std::string AsFortran() const {
+      std::string result;
+      {
+        llvm::raw_string_ostream bs{result};
+        if (lower) {
+          evaluate::Constant<T>{*lower}.AsFortran(bs << '(');
+          if (!upper) {
+            bs << ':';
+          } else if (*lower != *upper) {
+            evaluate::Constant<T>{*upper}.AsFortran(bs << ':');
+          }
+          bs << ')';
+        } else if (upper) {
+          evaluate::Constant<T>{*upper}.AsFortran(bs << "(:") << ')';
+        } else {
+          bs << "DEFAULT";
+        }
+      }
+      return result;
```

Comment:
You can avoid needing the extra block with `return bs.str();`

Reply:
As written, I think it saves a copy.